### PR TITLE
Move vector mesh generation to napari vispy visual and fix vectors refesh

### LINKED ISSUE #4694: Vectors layer not refreshing on refresh()
## 🐛 Bug

Vectors layer not refreshing despite having updated elements of the array.
EDIT: Vectors layer not refreshing on `refresh()`

## To Reproduce

```
import napari
import numpy as np
v = napari.Viewer()
data = np.array([[0,10],[0,5]])
viewer.add_vectors(data)
viewer.layers[0].data[0] = np.array([[0,10],[0,1]])
# Look at the data: it has been updated
viewer.layers[0].data
# But the viewer has not, event after triggering refresh
viewer.layers[0].refresh()
# But setting edge_width updates the viewer!
viewer.layers[0].edge_width=1
```

## Expected behavior

The viewer should update when changing parts of the array (because it does when you change the whole data at once) or at least the refresh() method should work...
EDIT: The viewer should update when triggering `refresh()`

## Environment
napari: 0.4.16
Platform: Linux-5.17.13-200.fc35.x86_64-x86_64-with-glibc2.34
System: Fedora Linux 35 (Workstation Edition)
Python: 3.9.13 (main, May 18 2022, 00:00:00) [GCC 11.3.1 20220421 (Red Hat 11.3.1-2)]
Qt: 5.15.2
PyQt5: 5.15.6
NumPy: 1.22.4
SciPy: 1.7.3
Dask: 2022.05.2
VisPy: 0.10.0

OpenGL:
- GL version: 4.6.0 NVIDIA 515.48.07
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1200, scale 1.0

Plugins:
- console: 0.0.4
- napari-deepfinder: 0.0.1
- napari-svg: 0.1.6
- scikit-image: 0.4.16

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR picks up on discussion https://github.com/napari/napari/issues/4694#issuecomment-1183794908 to explore moving mesh generation for the vectors layer to the vispy visual.

This approach has the advantage of consolidating complex mesh logic further away from the napari layer, but the disadvantage that mesh generation is now done each time the slice is changed. The mesh generation though now only uses data that is in the current slice though, instead of pre-generating the mesh for all possible slices (which could have been very expensive).

This also closes #4694 as now refresh will regenerate mesh - we should probably add a test for that before merge.

This was a fairly quick implementation, and so i may have missed something - maybe @brisvag @andy-sweet @kcpevey can look at/ take over and see if you like direction this is going. I ended up duplicating the `segment_normal` function to avoid some annoying circular imports.

## Type of change

- [x] Refactor

# References

https://github.com/napari/napari/issues/4694#issuecomment-1183794908

# How has this been tested?

- [x] using napari examples with vectors

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by brisvag** [comment]
> but the disadvantage that mesh generation is now done each time the slice is changed

This was the thing that got me stuck last time, as I was trying to work around it. However, now I tested this branch with some reasonably sized vector layers, and I didn't see a noticeable change in performance. I'm testing on a workstation, though, so maybe someone can take a look on a laptop?

**Comment by Czaki** [comment]
maybe there should be some caching for mesh?

**Comment by andy-sweet** [comment]
Code changes look great! Your answer to the problem of deciding when to regenerate meshes is always, which I also like because of its simplicity. But as you point out it does need some benchmarking and/or profiling. The existing vectors layer benchmarks don't help because they only benchmark the layer (e.g. `Vectors._set_view_slice`), which does less now.

Maybe we should define some more integrated benchmarks that include generating and pushing the right data to vispy?

**Comment by liu-ziyang** [comment]
@kcpevey do you mind taking this PR from here? I hope you have some capacity to help on this

**Comment by sofroniewn** [comment]
> Maybe we should define some more integrated benchmarks that include generating and pushing the right data to vispy?

Sounds good as a next step.

@brisvag @alisterburt do you have any real-world datasets we could try here. Maybe some people doing polarization microscopy might as well - cc @chili-chiu 

> @kcpevey do you mind taking this PR from here? I hope you have some capacity to help on this

That would be great!

**Comment by brisvag** [comment]
A good representative dataset for me is not that big: about 5k to 10k vectors in a 3D volume; I see no performance hit there. Maybe someone with heavier datasets can chip in :)

**Review by brisvag — APPROVED** [review]
Unless performance concerns arise, I would say this is a net improvement over main, so I'm approving!

**Comment by kcpevey** [comment]
>@kcpevey do you mind taking this PR from here? I hope you have some capacity to help on this

happy to! Just coming back from vacation, but I'm back now!

File: napari/layers/vectors/vectors.py [hunk 1]
```diff
@@ -656,48 +634,22 @@ def _set_view_slice(self):
...
             self._view_indices = np.arange(self.data.shape[0])
             self._view_alphas = 1.0
 
-        if len(faces) == 0:
-            self._view_vertices = []
-            self._view_faces = []
-        else:
-            self._view_vertices = vertices
-            self._view_faces = faces
-
```
Comment (kcpevey) [inline napari/layers/vectors/vectors.py:714]: I really like all this cleanup. I had done something similar for the async slicing vector prototype.

**Comment by sofroniewn** [comment]
> I really like all this cleanup. I had done something similar for the async slicing vector prototype.

Great @kcpevey! So i think all this PR needs to get over the finish line is a bit of performance testing. It would also be nice not to duplicate the segment_normal function - not sure how to fix the circular imports. We could also add some unit tests for `generate_vector_meshes` and `generate_vector_meshes_2D` that probably should have been added when they were originally contributed, but now will be as good a time as any. Let me know if you have any questions @kcpevey - you can either make PRs to this PR, or if you want to make your own PR to napari with these commits that's fine too - really whatever is easiest and fastest for you. Thanks again!!

**Comment by kcpevey** [comment]
@sofroniewn I spent some time today sorting through the best way to benchmark this. There is definitely some trickiness to getting the qt events to trigger inside of the benchmarking suite. I *think* I sorted out something effective [see here](https://github.com/sofroniewn/napari/pull/15). 

I will write some unit tests for the `generate_vector_meshes` functions too and take a look at `segment_normal` as well.

**Comment by kcpevey** [comment]
My continuation of this work is in this PR: https://github.com/sofroniewn/napari/pull/17

**Comment by sofroniewn** [comment]
Thanks to work from @kcpevey this PR is now ready for final review. @kcpevey has added some tests and benchmarks. @brisvag can you give it a final look and if you like it can you merge it? Thanks!!

See here for benchmark info

> I rechecked the benchmarks against main (where this branch started).
> 
> **main**
> 
> ```
> [ 75.00%] ··· ...iewer_vectors.QtViewerViewVectorSuite.time_vectors_multi_refresh                 
> [ 75.00%] ··· ======== =============
>                param1               
>               -------- -------------
>                  16     9.06±0.09ms 
>                  64       15.9±1ms  
>                 256      37.5±0.8ms 
>                 1024      127±2ms   
>                 4096      275±2ms   
>                16384      430±30ms  
>                65536      972±4ms   
>               ======== =============
> 
> [100.00%] ··· ...k_qt_viewer_vectors.QtViewerViewVectorSuite.time_vectors_refresh                
> [100.00%] ··· ======== =============
>                param1               
>               -------- -------------
>                  16     3.03±0.04ms 
>                  64      5.46±0.1ms 
>                 256     12.2±0.07ms 
>                 1024      43.6±2ms  
>                 4096      90.6±1ms  
>                16384      138±5ms   
>                65536      319±4ms   
>               ======== =============
> ```
> 
> **this branch**
> 
> ```
> [ 75.00%] ··· ...iewer_vectors.QtViewerViewVectorSuite.time_vectors_multi_refresh                
> [ 75.00%] ··· ======== ============
>                param1              
>               -------- ------------
>                  16     9.37±0.2ms 
>                  64     15.6±0.2ms 
>                 256     38.2±0.9ms 
>                 1024     130±2ms   
>                 4096     291±10ms  
>                16384     438±6ms   
>                65536    1.13±0.01s 
>               ======== ============
> 
> [100.00%] ··· ...k_qt_viewer_vectors.QtViewerViewVectorSuite.time_vectors_refresh                 
> [100.00%] ··· ======== =============
>                param1               
>               -------- -------------
>                  16     3.06±0.03ms 
>                  64     5.38±0.07ms 
>                 256      12.9±0.5ms 
>                 1024      43.4±2ms  
>                 4096      95.6±5ms  
>                16384     152±0.9ms  
>                65536      368±4ms   
>               ======== =============
> ```

**Comment by kcpevey** [comment]
@sofroniewn who else do we need for review?

**Comment by sofroniewn** [comment]
I think this is now good to go - thanks for work @kcpevey and original review @brisvag, i will merge!!

**Pull request merged**